src/swupd/fullfiles.go

```diff
@@ -0,0 +1,229 @@
+
+var fullfileCompressors = []struct {
+	Name string
+	Func fullfileCompressFunc
+}{
+	{"gzip", compressGzip},
+}
+
+func CreateFullfiles(m *Manifest, chrootDir, outputDir string) error {
+	done := make(map[hashval]bool)
+	for _, f := range m.Files {
+		if done[f.Hash] || f.Version != m.Header.Version || f.Status == statusDeleted || f.Status == statusGhosted {
+			continue
+		}
+
+		input := filepath.Join(chrootDir, f.Name)
+		name := f.Hash.String()
+		// NOTE: to make life simpler for the client, always use .tar extension even if the
+		// file could be compressed.
+		output := filepath.Join(outputDir, name+".tar")
+
+		var err error
+		switch f.Type {
+		case typeDirectory:
+			err = createDirectoryFullfile(input, name, output)
+		case typeLink:
+			err = createLinkFullfile(input, name, output)
+		case typeFile:
+			err = createRegularFullfile(input, name, output)
+		default:
+			return fmt.Errorf("cannot create fullfiles: file %s is of unsupported type %q", f.Name, f.Type)
+		}
+
+		if err != nil {
+			return err
+		}
+
+		done[f.Hash] = true
+	}
+	return nil
+}
+
+func createDirectoryFullfile(input, name, output string) error {
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if !fi.IsDir() {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a directory but it is not", input)
+	}
+
+	hdr, err := getHeaderFromFileInfo(fi)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	hdr.Name = name
+	hdr.Typeflag = tar.TypeDir
+
+	err = createTarGzipHeaderOnly(output, hdr)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	return nil
+}
+
+func createLinkFullfile(input, name, output string) error {
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if fi.Mode()&os.ModeSymlink == 0 {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a link but it is not", input)
+	}
+	link, err := os.Readlink(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	hdr, err := getHeaderFromFileInfo(fi)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	hdr.Name = name
+	hdr.Typeflag = tar.TypeSymlink
+	hdr.Linkname = link
+
+	err = createTarGzipHeaderOnly(output, hdr)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	return nil
+}
+
+func createRegularFullfile(input, name, output string) error {
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if !fi.Mode().IsRegular() {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a regular file but it is not", input)
+	}
+
+	best := ""
+	var bestSize int64 = math.MaxInt64
```

Comment:
It could work but at the moment if all our compressors give something larger than the original file, that would cause a failure that we don't handle yet.